src/dex/woo-fi/woo-fi.ts

```diff
@@ -0,0 +1,556 @@
+import _ from 'lodash';
+import { Interface, Result } from '@ethersproject/abi';
+import {
+  Token,
+  Address,
+  ExchangePrices,
+  AdapterExchangeParam,
+  SimpleExchangeParam,
+  PoolLiquidity,
+  Logger,
+} from '../../types';
+import { SwapSide, Network, NULL_ADDRESS } from '../../constants';
+import { getBigIntPow, getDexKeysWithNetwork, wrapETH } from '../../utils';
+import { IDex } from '../../dex/idex';
+import { IDexHelper } from '../../dex-helper/idex-helper';
+import { PoolState, WooFiData } from './types';
+import { SimpleExchange } from '../simple-exchange';
+import { WooFiConfig, Adapters } from './config';
+import { wooFiMath } from './woo-fi-math';
+import {
+  MIN_CONVERSION_RATE,
+  USD_PRECISION,
+  WOO_FI_GAS_COST,
+} from './constants';
+import wooPPABI from '../../abi/woo-fi/WooPP.abi.json';
+import wooFeeManagerABI from '../../abi/woo-fi/WooFeeManager.abi.json';
+import woOracleABI from '../../abi/woo-fi/Wooracle.abi.json';
+
+export class WooFi extends SimpleExchange implements IDex<WooFiData> {
+  readonly math: typeof wooFiMath = wooFiMath;
+
+  latestState: PoolState | null = null;
+
+  latestBlockNumber: number = 0;
+
+  tokenByAddress: Record<string, Token> | null = null;
+
+  readonly wooIfaces = {
+    PP: new Interface(wooPPABI),
+    fee: new Interface(wooFeeManagerABI),
+    oracle: new Interface(woOracleABI),
+  };
+
+  readonly hasConstantPriceLargeAmounts = false;
+  readonly needWrapNative = true;
+
+  private _exchangeAddress?: string;
+
+  private _oracleAddress?: string;
+
+  private _quoteTokenAddress?: string;
+
+  public static dexKeysWithNetwork: { key: string; networks: Network[] }[] =
+    getDexKeysWithNetwork(WooFiConfig);
+
+  logger: Logger;
+
+  constructor(
+    protected network: Network,
+    protected dexKey: string,
+    protected dexHelper: IDexHelper,
+    protected adapters = Adapters[network] || {},
+    readonly config = WooFiConfig[dexKey][network],
+  ) {
+    super(dexHelper.augustusAddress, dexHelper.provider);
+    this.logger = dexHelper.getLogger(dexKey);
+  }
+
+  get baseTokens(): Token[] {
+    return Object.values(this.config.baseTokens);
+  }
+
+  get exchangeAddress() {
+    if (this._exchangeAddress === undefined) {
+      this._exchangeAddress = this.config.wooPPAddress.toLowerCase();
+    }
+    return this._exchangeAddress;
+  }
+
+  get oracleAddress() {
+    if (this._oracleAddress === undefined) {
+      this._oracleAddress = this.config.woOracleAddress.toLowerCase();
+    }
+    return this._oracleAddress;
+  }
+
+  get quoteTokenAddress() {
+    if (this._quoteTokenAddress === undefined) {
+      this._quoteTokenAddress = this.config.quoteToken.address.toLowerCase();
+    }
+    return this._quoteTokenAddress;
+  }
```

Comment:
It would be cleaner to set `_exchangeAddress , _oracleAddress , _quoteTokenAddress` in the constructor instead of every time checking if its not defined